snoop/digest.py

```diff
@@ -100,65 +100,55 @@ def digest(doc):
-    elif pst.is_pst_file(doc):
-        data.update(pst.list_files(doc))
+        archives.extract_to_base(doc)
+    if pst.is_pst_file(doc):
+        pst.extract_to_base(doc)
 
     return data
 
-def create_children(doc, data, verbose=True):
-    children_info = []
-    if emails.is_email(doc):
-        for name, info in data.get('attachments', {}).items():
-            children_info.append({
-                'path': name,
+def create_email_children(doc, data, flags):
+    children = []
+    for name, info in data.get('attachments', {}).items():
+        child, created = models.Document.objects.update_or_create(
+            container=doc,
+            parent=doc,
+            path=name,
+            defaults={
+                'disk_size': info.get('size', 0),
                 'content_type': info['content_type'],
                 'filename': info['filename'],
-                'size': info.get('size', 0),
-            })
-    elif archives.is_archive(doc) or pst.is_pst_file(doc):
-        for path in data['file_list']:
-            children_info.append({
-                'path': path,
-                'content_type': guess_content_type(path),
-                'filename': Path(path).name,
-                'size': 0,
-            })
-        for path in data['folder_list']:
-            children_info.append({
-                'path': path,
-                'content_type': 'application/x-directory',
-                'filename': Path(path).name,
-                'size': 0,
-            })
+            },
+        )
+        children.append((child.id, created))
+    return children
 
+def create_children(doc, data, verbose=True):
     inherited_flags = {
         key: doc.flags[key]
         for key in doc.flags
         if key in INHERITABLE_DOCUMENT_FLAGS
     }
+    children = []
 
-    new_children = 0
-    for info in children_info:
-        child, created = models.Document.objects.update_or_create(
-            container=doc,
-            path=info['path'],
-            defaults={
-                'disk_size': info['size'],
-                'content_type': info['content_type'],
-                'filename': info['filename'],
-                'flags': inherited_flags,
-            },
-        )
+    if emails.is_email(doc):
+        children = create_email_children(doc, data, inherited_flags)
+    elif archives.is_archive(doc):
+        children = archives.list_children(doc)
+    elif pst.is_pst_file(doc):
+        children = pst.list_children(doc)
 
+    new_children = 0
+    for doc_id, created in children:
+        child = models.Document.objects.get(id=doc_id)
+        child.flags.update(inherited_flags)
+        child.save()
         if created:
-            queues.put('digest', {'id': child.id}, verbose=verbose)
-            if verbose: print('new child', child.id)
             new_children += 1
+            queues.put('digest', {'id': doc_id}, verbose=verbose)
+            if verbose: print('new child', doc_id)
```

Comment:
Do we need to swap these two lines with `new_children += 1`? If not, let's keep them as they were, so the diff is nicer to read :)